pkg/controlplane/controller/kubernetesservice/controller.go

```diff
@@ -81,11 +88,13 @@ func (c *Controller) Start(stopCh <-chan struct{}) {
 		klog.Errorf("Error removing old endpoints from kubernetes service: %v", err)
 	}
 
-	go c.Run(stopCh)
+	go c.Run(c.stopCh)
 }
 
 // Stop cleans up this API Servers endpoint reconciliation leases so another master can take over more quickly.
 func (c *Controller) Stop() {
+	close(c.stopCh)
```

Comment:
attempting to close a closed channel panics, same above

```
select {
case <-c.stopCh:
default:
   close(c.stopCh)
}
```

Reply:
maybe we can replace the channel with a context and share the cancel function? that can be executed multiple times.

there is a `wait.ContextForChannel`helper

Reply:
That races. Updated last commit. Ptal.

Reply:
It will also race storing the cancelFn in the controller struct :-) It don't think we get around any kind of synchroniziation (now with a select).

Reply:
would call this resolved.

Reply:
yeah, this is not worse than before and since this version the reconciler handles better the concurrency https://github.com/kubernetes/kubernetes/pull/116685

As TODO we need to make Stop wait for the Start to finish

Reply:
actually I think this is ok since Stop will calls `c.EndpointReconciler.StopReconciling()` and the reconciler will handle the race,